src/rgw/s3select.h

```diff
@@ -0,0 +1,852 @@
+#ifndef __S3SELECT__
+#define __S3SELECT__
+
+#include <boost/spirit/include/classic_core.hpp>
+#include <iostream>
+#include <string>
+#include <list>
+#include "s3select_oper.h"
+#include <boost/function.hpp>
+#include <boost/bind.hpp>
+#include <functional>
+
+using namespace std;
+using namespace BOOST_SPIRIT_CLASSIC_NS;
+#define _DEBUG_TERM {string  token(a,b);std::cout << __FUNCTION__ << token << std::endl;}
+
+/// AST builder
+
+class s3select_projections {
+
+    private:
+        list<base_statement *> m_projections;
+    
+    public:
+        bool is_aggregate()
+        {
+            //TODO iterate on projections , and search for aggregate
+            //for(auto p : m_projections){}
+
+	    return false;
+        }
+
+        bool semantic()
+        {
+            //TODO check aggragtion function are not nested 
+            return false;
+        }
+
+        list<base_statement *> * get()
+        {
+            return &m_projections;
+        }
+
+};
+
+struct actionQ
+{
+
+    list<mulldiv_operation::muldiv_t> muldivQ;
+    list<addsub_operation::addsub_op_t> addsubQ;
+    list<arithmetic_operand::cmp_t> arithmetic_compareQ;
+    list<logical_operand::oplog_t> logical_compareQ;
+    list<base_statement *> exprQ;
+    list<base_statement *> funcQ;
+    list<base_statement *> condQ;
+    std::string from_clause;
+    list<std::string> schema_columns;
+    s3select_projections  projections;
+
+};
+
+class base_action
+{
+
+public:
+    base_action() : m_action(0) {}
+    actionQ *m_action;
+    void set_action_q(actionQ *a) { m_action = a; }
+};
+
+struct push_from_clause : public base_action
+{
+    void operator()(const char *a, const char *b) const
+    {
+        std::string token(a, b);
+        m_action->from_clause = token;
+    }
+
+};
+static push_from_clause g_push_from_clause;
+
+struct push_number : public base_action //TODO use define for defintion of actions
+{
+    void operator()(const char *a, const char *b) const
+    {
+        string token(a, b);
+        variable *v = new variable(atoi(token.c_str())); //TODO strtoll
```

Comment:
I'll put this here, but this is more of a general request for comment;  the unique_ptr discussion has the corollary, could we more generally avoid heap allocation--potentially just relying on strings, std::move/move semantics, and some use of std::string_view (avoiding the whole issue of heap allocation) as a general strategy?

Reply:
sure , totally agreed with that.
indeed , the general strategy is to avoid heap allocation (placement new is one technique),
in this part (AST builder) it allocate once per query life cycle (parsing stage).

as mentioned in previous comment, the call-stack starting at /send_response_data/ should modified accordingly.